pkg/controller/session/session_controller_test.go

```diff
@@ -0,0 +1,321 @@
+			Revertors: []model.Revertor{revertor.Do},
+		}
+
+		schema, _ = v1alpha1.SchemeBuilder.Build()
+		req = reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      "test-session",
+				Namespace: "test",
+			},
+		}
+		c = fake.NewFakeClientWithScheme(schema, objects...)
+		controller = session.NewStandaloneReconciler(c, manipulators)
+	})
+
+	Context("session creation", func() {
+
+		Context("session not found", func() {
+			BeforeEach(func() {
+				objects = []runtime.Object{}
+			})
+
+			It("no retry", func() {
+				res, err := controller.Reconcile(req)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
+			})
+		})
+		Context("session found", func() {
+			BeforeEach(func() {
+				objects = []runtime.Object{
+					&v1alpha1.Session{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      "test-session",
+							Namespace: "test",
+						},
+						Spec: v1alpha1.SessionSpec{
+							Refs: []string{"details"},
+						},
+					},
+				}
+			})
+
+			It("finalizer added", func() {
+				res, err := controller.Reconcile(req)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
+
+				modified := GetSession("test", "test-session")
+				Expect(modified.ObjectMeta.Finalizers).To(HaveLen(1))
+			})
+
+			It("found locator", func() {
+				locator.Action = foundTestLocator
+
+				res, err := controller.Reconcile(req)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
+
+				Expect(locator.WasCalled).To(BeTrue())
+				Expect(mutator.WasCalled).To(BeTrue())
+			})
+
+			It("status created", func() {
+				locator.Action = foundTestLocator
+				mutator.Action = basicTestMutator(model.ResourceStatus{Name: "details", Kind: "test", Action: model.ActionCreated})
+
+				res, err := controller.Reconcile(req)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
+
+				Expect(locator.WasCalled).To(BeTrue())
+				Expect(mutator.WasCalled).To(BeTrue())
+				Expect(revertor.WasCalled).ToNot(BeTrue())
+
+				modified := GetSession("test", "test-session")
+				Expect(modified.Status).ToNot(BeNil())
+				Expect(modified.Status.Refs).To(HaveLen(1))
+				Expect(modified.Status.Refs[0].Name).To(Equal("details"))
+			})
+		})
+	})
+
+	Context("session modification", func() {
+		Context("new reference", func() {
+			BeforeEach(func() {
+				objects = []runtime.Object{
+					&v1alpha1.Session{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:       "test-session",
+							Namespace:  "test",
+							Finalizers: []string{session.Finalizer},
+						},
+						Spec: v1alpha1.SessionSpec{
+							Refs: []string{"details", "details2"},
+						},
+						Status: v1alpha1.SessionStatus{
+							Refs: []*v1alpha1.RefStatus{
+								{
+									Name:      "details",
+									Resources: []*v1alpha1.RefResource{{Kind: &kind, Name: &name, Action: &action}},
+								},
+							},
+						},
+					},
+				}
+			})
+			It("status updated", func() {
```

Comment:
How about making it a bit more elaborated - e.g. `It("should update status when .... ?"`

Reply:
https://github.com/aslakknutsen/istio-workspace/pull/90/commits/5c416866df4c6fc26d4058cba5ccf04ca2a64770